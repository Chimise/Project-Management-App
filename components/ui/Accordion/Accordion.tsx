import React, { useState } from "react";
import cn from 'classnames';
import { Disclosure, Transition } from "@headlessui/react";

type IconProps = React.ComponentProps<"svg"> & {
  title?: string;
  titleId?: string;
};

interface AccordionProps {
  openIcon: React.JSXElementConstructor<IconProps>;
  closeIcon?: React.JSXElementConstructor<IconProps>;
  title: string;
  children: React.ReactNode;
  onView: () => void;
  viewed: boolean;
}

const Accordion = ({
  openIcon: OpenIcon,
  closeIcon: CloseIcon,
  title,
  children,
  onView,
  viewed
}: AccordionProps) => {
  const Icon = CloseIcon || OpenIcon;
  const [isVisible, setIsVisible] = useState(false);

  return (
    <Disclosure data-viewed={viewed} as="div" className="w-full transition-all duration-200 hover:-translate-x-1 hover:bg-primary/10">
      {({ open }) => (
        <>
          <Disclosure.Button
            onMouseEnter={() => setIsVisible(true)}
            onMouseLeave={() => setIsVisible(false)}
            onClick={onView}
            className="flex items-center p-3 overflow-hidden w-full"
          >
            <div className="w-10 flex justify-center text-primary items-center">
              {viewed ? (
                <OpenIcon className="w-7 h-7 text-primary opacity-50" />
              ) : (
                <Icon className="w-7 h-7 text-primary" />
              )}
            </div>
            <div className="flex flex-1 px-3 items-center justify-between">
            <h6 className={cn('text-base md:text-lg text-black', {'font-medium': !viewed, 'font-normal': viewed})}>{title}</h6>
              <Transition
                show={isVisible}
                enter="transition-all duration-200"
                enterFrom="opacity-0 translate-x-full"
                enterTo="opacity-1 translate-x-0"
                leave="transition-all duration-200"
                leaveFrom="opacity-1 translate-x-0"
                leaveTo="opacity-0 translate-x-full"
              >
                <div  className="text-primary text-sm">Click to {open ? "Close" : "Open"}</div>
              </Transition>
            </div>
          </Disclosure.Button>
          <Disclosure.Panel className="w-full flex min-h-[40px] px-3 pb-3">
            <div className="w-10 flex justify-center items-stretch">
              <span className={cn("h-full w-1 bg-primary", {'opacity-50': viewed, 'opacity-100': !viewed})} />
            </div>
            <div className="flex-1 px-3">{children}</div>
          </Disclosure.Panel>
        </>
      )}
    </Disclosure>
  );
};

export default Accordion;
