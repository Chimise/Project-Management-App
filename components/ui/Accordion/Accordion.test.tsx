import "@testing-library/jest-dom";
import {
  render,
  user,
  waitForElementToBeRemoved
} from "../../../test/test-utils";
import router, { mockUseRouter, mockPush } from "../../../test/router";
import Accordion from "./Accordion";
import {EnvelopeIcon} from '@heroicons/react/20/solid'

beforeEach(() => {
  jest.useFakeTimers();
  jest.clearAllMocks();
});

afterEach(() => {
  jest.useRealTimers();
});

mockUseRouter();
const handler = jest.fn();
const children = <span data-testid='child' >This is only visible when the accordion is clicked</span>
const title = 'This is an accordion';
test('It renders unchanged', () => {
    const {container} = render(<Accordion openIcon={EnvelopeIcon} title={title} onView={handler} viewed={false} >{children}</Accordion>);
    expect(container.firstChild).toMatchSnapshot();
})

test('It shows a slide text on hover which disappears on unhover', async () => {
    const {getByRole, getByText} = render(<Accordion openIcon={EnvelopeIcon} title={title} onView={handler} viewed={false} >{children}</Accordion>);
    const accordion = getByRole('button');
    await user.hover(accordion);
    const slideText = getByText(/click to open/i);
    expect(slideText).toBeInTheDocument();
    user.unhover(slideText);
    await waitForElementToBeRemoved(slideText);
});

test('The child element displays when the accordion is clicked and the onView callback is also called', async () => {
    const {getByRole, queryByTestId, getByTestId} = render(<Accordion openIcon={EnvelopeIcon} title={title} onView={handler} viewed={false} >{children}</Accordion>);
    const accordion = getByRole('button');
    expect(queryByTestId('child')).not.toBeInTheDocument();
    await user.click(accordion);
    const child = getByTestId('child');
    expect(child).toBeInTheDocument();
    expect(handler).toHaveBeenCalledTimes(1);
    user.click(accordion);
    await waitForElementToBeRemoved(child);
});