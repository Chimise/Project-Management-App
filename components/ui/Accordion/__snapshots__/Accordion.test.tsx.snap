// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It renders unchanged 1`] = `
<div
  class="w-full transition-all duration-200 hover:-translate-x-1 hover:bg-primary/10"
  data-headlessui-state=""
  data-viewed="false"
>
  <button
    aria-expanded="false"
    class="flex items-center p-3 overflow-hidden w-full"
    data-headlessui-state=""
    id="headlessui-disclosure-button-:r1:"
    type="button"
  >
    <div
      class="w-10 flex justify-center text-primary items-center"
    >
      <svg
        aria-hidden="true"
        class="w-7 h-7 text-primary"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 4a2 2 0 00-2 2v1.161l8.441 4.221a1.25 1.25 0 001.118 0L19 7.162V6a2 2 0 00-2-2H3z"
        />
        <path
          d="M19 8.839l-7.77 3.885a2.75 2.75 0 01-2.46 0L1 8.839V14a2 2 0 002 2h14a2 2 0 002-2V8.839z"
        />
      </svg>
    </div>
    <div
      class="flex flex-1 px-3 items-center justify-between"
    >
      <h6
        class="text-base truncate md:text-lg text-black font-medium"
      >
        This is an accordion
      </h6>
    </div>
  </button>
</div>
`;
